admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -2,66 +2,115 @@
+    mempool::{AddTransactionWithValidationRequest, HealthCheckRequest},
+    mempool_client::MempoolClientTrait,
+};
+use libra_mempool_shared_proto::proto::mempool_status::{
+    MempoolAddTransactionStatus,
+    MempoolAddTransactionStatusCode::{self, MempoolIsFull},
+};
 use libra_prost_ext::MessageExt;
+use libra_types::transaction::SignedTransaction;
 use network::validator_network::{
     AdmissionControlNetworkEvents, AdmissionControlNetworkSender, Event, RpcError,
 };
 use rand::seq::SliceRandom;
 use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::sync::Arc;
+use storage_client::StorageRead;
 use tokio::runtime::TaskExecutor;
+use vm_validator::vm_validator::{get_account_state, TransactionValidation};
+
+#[cfg(test)]
+#[path = "unit_tests/admission_control_service_test.rs"]
+mod admission_control_service_test;
+
+#[cfg(feature = "fuzzing")]
+#[path = "admission_control_fuzzing.rs"]
+/// fuzzing module for admission control
+pub mod fuzzing;
 
 /// UpstreamProxyData is the set of data needed for a full node to send transaction write
 /// requests to their upstream validator.
 /// UpstreamProxyData is instantiated in AC Runtime.
 #[derive(Clone)]
-pub struct UpstreamProxyData {
+pub struct UpstreamProxyData<M, V> {
     ac_config: AdmissionControlConfig,
     network_sender: AdmissionControlNetworkSender,
     role: RoleType,
-    /// AC Client
-    pub client: AdmissionControlClient, // TODO remove client
+    /// gRPC client connecting Mempool.
+    mempool_client: Option<Arc<M>>,
+    /// gRPC client to send read requests to Storage.
+    storage_read_client: Arc<dyn StorageRead>,
+    /// VM validator instance to validate transactions sent from wallets.
+    vm_validator: Arc<V>,
+    /// Flag indicating whether we need to check mempool before validation, drop txn if check
+    /// fails.
+    need_to_check_mempool_before_validation: bool,
 }
 
-impl UpstreamProxyData {
+impl<M: 'static, V> UpstreamProxyData<M, V>
+where
+    M: MempoolClientTrait,
+    V: TransactionValidation,
+{
+    /// Render upstream proxy data
     pub fn new(
         ac_config: AdmissionControlConfig,
         network_sender: AdmissionControlNetworkSender,
         role: RoleType,
-        client: AdmissionControlClient, // TODO remove client
+        mempool_client: Option<Arc<M>>,
+        storage_read_client: Arc<dyn StorageRead>,
+        vm_validator: Arc<V>,
+        need_to_check_mempool_before_validation: bool,
     ) -> Self {
         Self {
             ac_config,
             network_sender,
             role,
-            client,
+            mempool_client,
+            storage_read_client,
+            vm_validator,
+            need_to_check_mempool_before_validation,
         }
     }
 }
 
 /// Main routine for proxying write request. Starts a coordinator that listens for AdmissionControlMsg
-pub async fn process_network_messages(
-    upstream_proxy_data: UpstreamProxyData,
+#[allow(clippy::implicit_hasher)]
```

Comment:
why do we need it?

Reply:
Because of this clippy warning: 
`error: parameter of type HashMap should be generalized over different hashers
  --> admission_control/admission-control-service/src/upstream_proxy.rs:103:20
   |
103 |     mut peer_info: HashMap<PeerId, bool>,
   |                    ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: -D clippy::implicit-hasher implied by -D warnings
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implicit_hasher
help: consider adding a type parameter`